service/controller/resource/alerting/heartbeatwebhookconfig/resource.go

```diff
@@ -0,0 +1,161 @@
+package heartbeatwebhookconfig
+
+import (
+	"fmt"
+	"net/url"
+	"reflect"
+	"strings"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	monitoringv1alpha1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1alpha1"
+	monitoringclient "github.com/prometheus-operator/prometheus-operator/pkg/client/versioned"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/prometheus-meta-operator/service/controller/resource/generic"
+	"github.com/giantswarm/prometheus-meta-operator/service/key"
+)
+
+const (
+	Name = "alertmanagerrouting"
+)
+
+type Config struct {
+	Client monitoringclient.Interface
+	Logger micrologger.Logger
+
+	Installation string
+	HTTPProxy    string
+	HTTPSProxy   string
+	NoProxy      string
+}
+
+func New(config Config) (*generic.Resource, error) {
+	clientFunc := func(namespace string) generic.Interface {
+		c := config.Client.MonitoringV1alpha1().AlertmanagerConfigs(namespace)
+		return wrappedClient{client: c}
+	}
+
+	c := generic.Config{
+		ClientFunc: clientFunc,
+		Logger:     config.Logger,
+		Name:       Name,
+		GetObjectMeta: func(v interface{}) (metav1.ObjectMeta, error) {
+			return getObjectMeta(v, config)
+		},
+		GetDesiredObject: func(v interface{}) (metav1.Object, error) {
+			return toAlertmanagerConfig(v, config)
+		},
+		HasChangedFunc: hasChanged,
+	}
+	r, err := generic.New(c)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return r, nil
+}
+
+func getObjectMeta(v interface{}, config Config) (metav1.ObjectMeta, error) {
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return metav1.ObjectMeta{}, microerror.Mask(err)
+	}
+
+	return metav1.ObjectMeta{
+		Name:      key.ClusterID(cluster),
+		Namespace: key.NamespaceMonitoring(),
+		Labels:    key.AlertmanagerLabels(),
+	}, nil
+}
+
+func toAlertmanagerConfig(v interface{}, config Config) (metav1.Object, error) {
+	if v == nil {
+		return nil, nil
+	}
+
+	objectMeta, err := getObjectMeta(v, config)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	labels := make(map[string]string)
+	for k, v := range key.AlertmanagerLabels() {
+		labels[k] = v
+	}
+
+	sendResolved := false
+	urlAddress, err := url.Parse(fmt.Sprintf("https://api.opsgenie.com/v2/heartbeats/%s/ping", key.HeartbeatName(cluster, config.Installation)))
```

Comment:
Please move this into a key package function

Reply:
Solved